torch/_inductor/fx_passes/post_grad.py

```diff
@@ -1518,6 +1522,38 @@ def allow_cpu_device(self, node: fx.Node) -> bool:
             torch.ops.aten.slice_scatter.default,
         )
 
+    def is_on_target_device(self, node: fx.Node) -> bool:
+        """
+        Returns whether a node is on the target device.
+        """
+        node_device = self.get_node_device(node)
+        return node_device is not None and node_device.type == self.target
+
+    def is_cpu_scalar_tensor(self, node: fx.Node) -> bool:
+        """
+        Returns whether a node is a cpu scalar tensor.
+        """
+        device = self.get_node_device(node)
+        is_cpu = device is not None and device.type == "cpu"
```

Comment:
Should we check that this node doesn't have any other device_put nodes ?

Reply:
Yes. I added a noop elimination for `device_put` if other `device_put` exists. We cannot just use the other `device_put` directly since it may appear at the end of the graph. fx graph does not support moving a node earlier. Using topo-sort seems too much here. Also, post_grad noop_elimination pass happens before `move_constructors_to_gpu` pass.